File: src/typical90/typical90_cc/main.jl
function cumsum_2d(mat::Matrix{T})::Matrix where {T}
    s = zeros(T, size(mat)[1] + 1, size(mat)[2] + 1)
    for y in 1:size(mat)[1]
        for x in 1:size(mat)[2]
            s[x + 1, y + 1] = s[x, y + 1] + s[x + 1, y] - s[x, y] + mat[x, y]
        end
    end
    return s
end

function sum(cumsum::Matrix{T}, s::Tuple{Int, Int}, e::Tuple{Int, Int})::Int where {T}
    return cumsum[e[1], e[2]] - cumsum[s[1], e[2]] - cumsum[e[1], s[2]] + cumsum[s[1], s[2]]
end

function main()
    n, k = readline() |> split .|> s -> parse(Int, s)
    siz = 5000
    d::Matrix{Int} = zeros(siz, siz)
    for i in 1:n
        a, b = readline() |> split .|> s -> parse(Int, s)
        d[a, b] += 1
    end

    cs = cumsum_2d(d)

    ans = 0
    for a in 1:siz - k
        for b in 1:siz - k
            ans = max(ans, sum(cs, (a, b), (a + k + 1, b + k + 1)))
        end
    end

    println(ans)
end

main()

File: src/typical90/typical90_cc/main.py


# 2D cumsum
class CumSum2D:
    def __init__(self, grid: list) -> None:
        w, h = len(grid[0]), len(grid)
        self.s = [[0 for i in range(w + 1)] for j in range(h + 1)]
        for i in range(h):
            for j in range(w):
                self.s[i + 1][j + 1] = self.s[i][j + 1] + self.s[i + 1][j] - self.s[i][j] + grid[i][j]

    def sum(self, s: tuple, e: tuple) -> int:
        """ [sx, ex) * [sy, ey) """
        return self.s[e[0]][e[1]] - self.s[s[0]][e[1]] - self.s[e[0]][s[1]] + self.s[s[0]][s[1]]

if __name__ == "__main__":
    n, k = list(map(int, input().split()))
    r = 5000
    mp = [[0 for i in range(r)] for j in range(r)]
    for i in range(n):
        a, b = list(map(int, input().split()))
        mp[a - 1][b - 1] = 1
    c = CumSum2D(mp)

    ans = 0
    for a in range(k, r):
        for b in range(k, r):
            ans = max(ans, c.sum((a - k - 1, b - k - 1), (a, b)))

    print(ans)
